src/login/login-activity/service.ts

```diff
@@ -0,0 +1,65 @@
+import { UserLoginActivityRecord } from 'knex/types/tables.js';
+import db from '../../database.js';
+import { User } from '../../types.js';
+
+const MAX_FAILED_ATTEMPTS = 5;
+
+export function reachedMaxAttempts(attempts: number) {
+  return attempts === MAX_FAILED_ATTEMPTS;
+}
+
+async function getLoginActivity(user: User): Promise<UserLoginActivityRecord | undefined> {
```

Comment:
Instead of returning `undefined`, how about always returning a record? Because maybe if there's no database record yet, you can guess what the record should be for the initial state